text/3898-thin-pointers-with-inline-metadata.md

```diff
@@ -0,0 +1,370 @@
+* [`thin_trait_object`], an attribute macro that makes a thin trait object (by manually
+  constructing the vtable).
+
+Now they can be rewritten as:
+- `List<T>` -> `&Thin<[T]>`
+- `ThinVec<T>`, technically `Box<(usize, Thin<[MaybeUninit<T>]>)>` (in representation)
+- `ThinBox<T>` -> `Box<Thin<T>>`
+- `BoxedTrait` -> `Box<Thin<dyn Trait>>`
+
+where much less boilerplate code is needed.
+
+[`List<T>`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/type.List.html
+[`ThinVec<T>`]: https://docs.rs/thin-vec/latest/thin_vec/struct.ThinVec.html
+[`ThinBox<T>`]: https://doc.rust-lang.org/std/boxed/struct.ThinBox.html
+[`thin_trait_object`]: https://docs.rs/thin_trait_object/latest/thin_trait_object/attr.thin_trait_object.html
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+## Add `ValueSized` to the sized hierarchy
+
+Regarding [sized hierarchy], `Thin` is more than `PointeeSized` but not `MetaSized`:
+- it is not `MetaSized` because the metadata is not carried by the pointer itself;
+- it is more than `PointeeSized` because we actually know its size by reading the metadata
+  stored inside.
+
+We need to add new stuff to the sized hierarchy, named `ValueSized`, to indicate a value of
+which the size is known by reading its value, as mentioned in [RFC 3729 (comments)].
+
+```rust
+// mod core::marker;
+
+/// Indicates that a type's size is known from reading its value.
+/// 
+/// Different from `MetaSized`, this requires pointer dereferences.
+#[lang_item = "value_sized"]
+pub trait ValueSized: PointeeSized {}
+
+// Change the bound of `MetaSized: PointeeSized` to `MetaSized: ValueSized`
+#[lang_item = "meta_sized"]
+pub trait MetaSized: ValueSized {}
+```
+
+For `dyn Trait + ValueSized` types, the `MetadataSize` entry of the common vtable entries
+will be a method of `fn size(&self) -> usize` which computes the value's size at runtime,
+instead of a compile-time constant value.
+
+
+[sized hierarchy]: https://github.com/rust-lang/rust/issues/144404
+[RFC 3729 (comments)]: https://github.com/davidtwco/rfcs/blob/sized-hierarchy/text/3729-sized-hierarchy.md#user-content-fn-9-ebb4bca70c758b473dca6f49c3bb3cbc
+
+## Public APIs
+
+The public APIs of `Thin` consist of 2 parts.
+
+### `Thin<T, U>`
+`Thin<T, U>` is a (maybe unsized) value of `T` with the metadata type of `U` carried on.
+
+Typically, `U = T` or `U` is some type that `T: Unsize<U>`.
+```rust
+// mod core::thin;
+
+/// Wrapping a DST `T` with its metadata inlined,
+/// then the pointers of `Thin<T>` are thin.
+///
+/// The generic type `U` is for two-stage construction of
+/// `Thin`, i.e., `Thin<T, U> where T: Unsize<U>` must be
+/// constructed first, then coerced (unsized) to `Thin<U>`
+/// (aka `Thin<U, U>`)
+#[repr(C)]
+pub struct Thin<T: Pointee, U: Pointee = T> {
+    metadata: U::Metadata,
+    data: EraseMetadata<T>,
+}
+
+// The size is known via reading its metadata.
+impl<U: Pointee> ValueSized for Thin<U> {}
```

Comment:
I think this requires `U: MetaSized`: there are `!MetaSized` types where you can't work out their size. You plausibly could get away with `U: ValueSized` but it would have weird/surprising semantics - I think it would ignore the metadata and call the value's `size` function.